sdk/search/azure-search-documents/azure/search/documents/_internal/_search_index_document_batching_client.py

```diff
@@ -0,0 +1,280 @@
+    :keyword str api_version: The Search API version to use for requests.
+    """
+    # pylint: disable=too-many-instance-attributes
+    _ODATA_ACCEPT = "application/json;odata.metadata=none"  # type: str
+    _DEFAULT_WINDOW = 0
+    _DEFAULT_BATCH_SIZE = 1000
+
+    def __init__(self, endpoint, index_name, credential, **kwargs):
+        # type: (str, str, AzureKeyCredential, **Any) -> None
+
+        api_version = kwargs.pop('api_version', None)
+        validate_api_version(api_version)
+        self._batch_size = kwargs.pop('batch_size', self._DEFAULT_BATCH_SIZE)
+        self._window = kwargs.pop('window', self._DEFAULT_WINDOW)
+        self._auto_flush = self._window > 0
+        self._index_documents_batch = IndexDocumentsBatch()
+        self._endpoint = endpoint  # type: str
+        self._index_name = index_name  # type: str
+        self._index_key = None
+        self._credential = credential  # type: AzureKeyCredential
+        self._client = SearchIndexClient(
+            endpoint=endpoint, index_name=index_name, sdk_moniker=SDK_MONIKER, **kwargs
+        )  # type: SearchIndexClient
+        self._timer = threading.Timer(self._window, self.flush)
+        if self._auto_flush:
+            self._timer.start()
+        self._persistence = kwargs.pop('persistence', None)
+
+    def cleanup(self):
+        # type: () -> None
+        if self._auto_flush:
+            self._timer.cancel()
+
+    def __repr__(self):
+        # type: () -> str
+        return "<SearchClient [endpoint={}, index={}]>".format(
+            repr(self._endpoint), repr(self._index_name)
+        )[:1024]
+
+    @property
+    def actions(self):
+        # type: () -> List[IndexAction]
+        """The list of currently index actions in queue to index.
+
+        :rtype: List[IndexAction]
+        """
+        return self._index_documents_batch.actions
+
+    @property
+    def succeeded_actions(self):
+        # type: () -> List[IndexAction]
+        """The list of currently succeeded index actions in queue.
+
+        :rtype: List[IndexAction]
+        """
+        return self._index_documents_batch.succeeded_actions
+
+    @property
+    def failed_actions(self):
+        # type: () -> List[IndexAction]
+        """The list of currently failed index actions in queue.
+
+        :rtype: List[IndexAction]
+        """
+        return self._index_documents_batch.failed_actions
+
+    @property
+    def batch_size(self):
+        # type: () -> int
+        return self._batch_size
+
+    def close(self):
+        # type: () -> None
+        """Close the :class:`~azure.search.SearchClient` session.
+
+        """
+        return self._client.close()
+
+    def flush(self, raise_error=False):
+        # type: (bool) -> None
+        """Flush the batch.
+
+        :param bool raise_error: raise error if there are failures during flushing
+            Default to False which re-queue the failed tasks and retry on next flush.
```

Comment:
Do we retry indefinitely, or is there a condition to raise even if raise_error is False (after too many retries for instance)

Reply:
We discussed before and for this release, we will always retry it if it is retry-able. 

If it is not retry-able, we can get the information from response and will put those actions into the failed queue.